### PR TITLE
P0896R4 stream iterator changes

### PR DESCRIPTION
**Pull request by CaseyCarter**
Nothing exotic here, really, just adding default constructors and (for the output iterators) difference types so the stream iterators properly model the C++20 iterator concepts. 

I _did_ strengthen some `noexcept` on my way through, and fixed what appears to be a longstanding bug: `ostream_iterator`'s data members should be `private` instead of `protected`.

Partially addresses #39.

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
The tests are failing with:
```
Failed Tests (4):
1:   libc++ :: std/iterators/stream.iterators/ostreambuf.iterator/types.pass.cpp:1
1:   libc++ :: std/iterators/stream.iterators/ostream.iterator/types.pass.cpp:1
1:   libc++ :: std/iterators/stream.iterators/ostreambuf.iterator/types.pass.cpp:0
1:   libc++ :: std/iterators/stream.iterators/ostream.iterator/types.pass.cpp:0
```

**Comment by CaseyCarter** [comment]
> The tests are failing with:
> 
> ```
> Failed Tests (4):
> 1:   libc++ :: std/iterators/stream.iterators/ostreambuf.iterator/types.pass.cpp:1
> 1:   libc++ :: std/iterators/stream.iterators/ostream.iterator/types.pass.cpp:1
> 1:   libc++ :: std/iterators/stream.iterators/ostreambuf.iterator/types.pass.cpp:0
> 1:   libc++ :: std/iterators/stream.iterators/ostream.iterator/types.pass.cpp:0
> ```

XFAILed here, and fix submitted upstream (https://reviews.llvm.org/D87459) just in case we can someday update libcxx from upstream again ;)

File: stl/inc/iterator [hunk 1]
```diff
@@ -350,30 +366,33 @@ public:
...
 
     istreambuf_iterator(istream_type& _Istr) noexcept : _Strbuf(_Istr.rdbuf()), _Got(!_Strbuf), _Val() {}
 
+    istreambuf_iterator(streambuf_type* _Sb) noexcept : _Strbuf(_Sb), _Got(!_Sb), _Val() {}
+
 private:
     class _Istreambuf_proxy {
     public:
-        _NODISCARD _Elem operator*() const {
+        _NODISCARD _Elem operator*() const noexcept(is_nothrow_copy_constructible_v<_Elem>) /* strengthened */ {
```
Comment (StephanTLavavej) [inline stl/inc/iterator:387]: I am confused by the exception specifications in `istreambuf_iterator` and its proxy. `istreambuf_iterator` stores `_Elem _Val` (not mentioned as an exposition-only data member). The Standard depicts `istreambuf_iterator` as having a `noexcept` default constructor *and* a `noexcept` copy constructor. This seems to suggest that `_Elem` must be nothrow default constructible and nothrow copy constructible. Yet here, you're querying `is_nothrow_copy_constructible_v<_Elem>`.

Should these operations be assumed to be nothrow because this is a "character type"? (I believe the answer is yes.)

File: stl/inc/iterator [hunk 2]
```diff
@@ -324,21 +340,21 @@ public:
...
 
-protected:
-    const _Elem* _Mydelim; // pointer to delimiter string (NB: not freed)
-    ostream_type* _Myostr; // pointer to output stream
+private:
+    const _Elem* _Mydelim = nullptr; // pointer to delimiter string (NB: not freed)
+    ostream_type* _Myostr = nullptr; // pointer to output stream
 };
 
 // CLASS TEMPLATE istreambuf_iterator
```
Comment (StephanTLavavej) [inline stl/inc/iterator:360]: Nitpick: It seems inconsistent that the `// wrap stream buffer as input iterator` comment is being retained for `istreambuf_iterator` (and similarly for `ostreambuf_iterator`, yet the comments are being removed for `istream_iterator` and `ostream_iterator`). It seems like they should be consistently removed.

File: tests/std/tests/P0896R4_stream_iterators/test.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+#include <iterator>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+using namespace std;
+
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
+
+template <class T, class CharT = char, class Traits = char_traits<CharT>, class Diff = ptrdiff_t>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_stream_iterators/test.cpp:17]: Technically need to include `<cstddef>` for `ptrdiff_t`.

File: tests/std/tests/P0896R4_stream_iterators/test.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    test_ostream<int>(cout);
+    test_ostream<string>(cout);
+    test_ostream<int>(wcout);
+    test_ostream<string>(wcout);
+
+    test_istreambuf<char>();
+    test_istreambuf<wchar_t>();
+
+    test_ostreambuf(cout);
+    test_ostreambuf(wcout);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_stream_iterators/test.cpp:124]: Naming nitpick: I find `test_istream` slightly confusing because you're testing `istream_iterator`, a different type. I'd recommend appending `_iterator` to these function names.

**Review by StephanTLavavej — APPROVED** [review]
Approved with a handful of nitpicks and one significant question about `noexcept` strengthening. Thanks!

File: stl/inc/iterator [hunk 5]
```diff
@@ -288,31 +298,37 @@ private:
...
 public:
     using iterator_category = output_iterator_tag;
     using value_type        = void;
-    using difference_type   = void;
-    using pointer           = void;
-    using reference         = void;
-
+#ifdef __cpp_lib_concepts
+    using difference_type = ptrdiff_t;
+#else
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/iterator:319]: I'm inclined toward "neither" in both places per the "rule of 5" exception; I don't see these conditionals growing in scope.

File: stl/inc/iterator (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/iterator:387]: The Standard is murky here since it doesn't depict `istreambuf_iterator` as having an exposition-only `charT` member: it's possible to implement `istreambuf_iterator` by grabbing a fresh character with `sbuf_->sgetc()` on each call to `operator*` and not memoizing the result within the `istreambuf_iterator` itself. Our justification for `noexcept` here relies on the fact that `istreambuf_iterator<charT, Traits>` uses `istream<charT, Traits>`, and `istream`'s first template argument must "represent a member of the set of types containing `char`, `wchar_t`, and any other implementation-defined character types that meet the requirements for a character on which any of the iostream components can be instantiated." per one of the most underspecified bits of standardese [iostreams.limits.pos]/2.

"the requirements for a character on which any of the iostream components can be instantiated" is nearly circular, and AFAICT effectively means "types an implementation supports here which we don't require implementations to document." It's far from explicit that such types must be trivially copyable and literal as needed to support [istreambuf.iterator.general]/1 "All specializations of `istreambuf_iterator` shall have a trivial copy constructor, a `constexpr` default constructor, and a trivial destructor" if stored as a member of `istreambuf_iterator`. This is far beyond the threshold of vagueness where I expect misuse in the wild. 

Altogether the result is that I'm being cautious with `noexcept` strengthening here. I'd be happier making these unconditional if we were to `static_assert` that `charT` is trivially copyable and nothrow-default-constructible in the body of `istreambuf_iterator`, but I don't know that it's worth the potential breakage of code that's misusing this iterator in the wild.

File: stl/inc/iterator (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/iterator:360]: The `[io]stream_iterator` comments were ever-so-slightly more offensive due to lack of punctuation making them hard to parse; I'll remove the comments from the `buf` flavors as well.

File: tests/std/tests/P0896R4_stream_iterators/test.cpp (hunk 4 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_stream_iterators/test.cpp:124]: I'm relying on the context added by the test name, which seems incredibly fragile.

File: stl/inc/iterator (hunk 1 above)
Comment (StephanTLavavej) [reply stl/inc/iterator:388]: Cautious is fine, thanks.

**Review by StephanTLavavej — APPROVED** [review]

File: tests/std/tests/P0896R4_stream_iterators/test.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    STATIC_ASSERT(output_iterator<I, const T&>);
+
+    STATIC_ASSERT(is_same_v<typename I::difference_type, ptrdiff_t>);
+    { [[maybe_unused]] constexpr I constexprConstructed{}; }
+    STATIC_ASSERT(is_nothrow_default_constructible_v<I>);
+#endif // __cpp_lib_concepts
+}
+
+template <class CharT = char, class Traits = char_traits<CharT>>
+void test_istreambuf_iterator() {
```
Comment (mnatsuhara) [inline tests/std/tests/P0896R4_stream_iterators/test.cpp:74]: Is this missing a test for the `noexcept`ness of `operator*`?

**Review by mnatsuhara — APPROVED** [review]
Just one question on test coverage, otherwise looks great! :)

File: tests/std/tests/P0896R4_stream_iterators/test.cpp (hunk 6 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_stream_iterators/test.cpp:74]: `istreambuf_iterator::operator*` is not `noexcept`, and can't be because our implementation lazily updates `_Val` in `operator*`. (Our mutable members and changes in `const` members are [100% non-conforming](http://eel.is/c++draft/res.on.data.races#3) and can't be fixed because ABI.)

**Comment by CaseyCarter** [comment]
We love iterator changes because we're not iter-haters! ❤️
